src/wal/src/local_storage_impl/record_encoding.rs

```diff
@@ -0,0 +1,276 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use bytes_ext::{Buf, BufMut, SafeBuf, SafeBufMut};
+use codec::Encoder;
+use generic_error::GenericError;
+use macros::define_result;
+use snafu::{ensure, Backtrace, ResultExt, Snafu};
+
+pub const RECORD_ENCODING_V0: u8 = 0;
+pub const NEWEST_RECORD_ENCODING_VERSION: u8 = RECORD_ENCODING_V0;
+
+pub const VERSION_SIZE: usize = 1;
+pub const CRC_SIZE: usize = 4;
+pub const RECORD_LENGTH_SIZE: usize = 4;
+pub const KEY_LENGTH_SIZE: usize = 4;
+pub const VALUE_LENGTH_SIZE: usize = 4;
+
+#[derive(Debug, Snafu)]
+pub enum Error {
+    #[snafu(display("Version mismatch, expect:{}, given:{}", expected, given))]
+    Version { expected: u8, given: u8 },
+
+    #[snafu(display("Failed to encode, err:{}", source))]
+    Encoding { source: bytes_ext::Error },
+
+    #[snafu(display("Failed to decode, err:{}", source))]
+    Decoding { source: bytes_ext::Error },
+
+    #[snafu(display("Invalid record: {}, backtrace:\n{}", source, backtrace))]
+    InvalidRecord {
+        source: GenericError,
+        backtrace: Backtrace,
+    },
+
+    #[snafu(display("Length mismatch: expected {} but found {}", expected, actual))]
+    LengthMismatch { expected: usize, actual: usize },
+
+    #[snafu(display("Checksum mismatch: expected {}, but got {}", expected, actual))]
+    ChecksumMismatch { expected: u32, actual: u32 },
+}
+
+define_result!(Error);
+
+/// Record format:
+///
+/// ```text
+/// +---------+--------+--------+------------+-----+--------------+-------+
+/// | version |  crc   | length | key length | key | value length | value |
+/// |  (u8)   | (u32)  | (u32)  |   (u32)    |     |     (u32)    |       |
+/// +---------+--------+--------+------------+-----+--------------+-------+
+/// ```
+pub struct Record<'a> {
+    /// The version number of the record.
+    pub version: u8,
+
+    /// The CRC checksum of the record.
+    pub crc: u32,
+
+    /// The length of the record (excluding version, crc and length).
+    pub length: u32,
+
+    /// The length of the key in bytes.
+    pub key_length: u32,
```

Comment:
Our key is fixed sized now(`table_id+seq_num`), so I think we don't need key length field.